File: app/ui/src/pages/pages.ts
/**
 * Pages. Doc 16 section 3.7's rail item and phase 12c.
 *
 * "Explorer, editor with write and preview, backlinks panel, unresolved link
 * creation." Two states in one view: a list of what the vault holds, and one
 * page open for reading and writing.
 *
 * The preview is markdown rendered to the same shapes an answer uses, with one
 * addition: a wikilink is a control rather than text. A resolved one opens the
 * page it names; an unresolved one offers to write it, which is doc 16 section
 * 3.1's "creates the page on click" and the reason an unresolved link is kept
 * rather than dropped.
 */

import { esc } from '../canvas/visual.js';
import type { PageDetail, PageRow } from '../rpc.js';
import { COPY } from '../strings.js';
import { ago, emptyState } from './shared.js';

export interface PagesState {
  /** The page being read, or `null` for the explorer. */
  open: PageDetail | null;
  /** Whether the open page is being written rather than read. */
  editing: boolean;
}

export function pagesToolsHTML(state: PagesState): string {
  if (!state.open) {
    return `<div class="seg"><button data-page-act="new">${COPY.pagesNew}</button></div>`;
  }
  return (
    `<div class="seg">` +
    `<button data-page-act="close">${COPY.pagesBack}</button>` +
    `<button data-page-act="${state.editing ? 'preview' : 'edit'}"${state.editing ? '' : ' class="on"'}>` +
    `${state.editing ? COPY.pagesPreview : COPY.pagesEdit}</button>` +
    `</div>`
  );
}

function row(p: PageRow): string {
  // Doc 16 section 3.2: a page saved from a card carries the card's citations,
  // and that is worth seeing without opening it, because it is the difference
  // between a page that can support a claim and one that is only context.
  // A count of zero is not worth a chip that reads like one: a card that cited
  // nothing makes a page that carries nothing, and saying "0 carried citations"
  // states the same thing twice while looking like a measurement.
  const carried = !p.from_card
    ? ''
    : p.citations_carried > 0
      ? `<span class="chip">${p.citations_carried} ${COPY.pagesCarried}</span>`
      : `<span class="chip">${COPY.pagesFromCard}</span>`;
  return (
    `<li class="lib-row" data-page="${esc(p.id)}">` +
    `<div class="what"><div class="line">` +
    `<span class="title">${esc(p.title)}</span>${carried}</div>` +
    `<p class="meta">${esc(p.file_path)}, ${COPY.pagesEdited} ${esc(ago(p.updated_at))}</p>` +
    `</div>` +
    `<div class="verbs">` +
    `<button data-page-act="open">${COPY.pagesOpen}</button>` +
    `<button class="danger" data-page-act="remove">${COPY.pagesRemove}</button>` +
    `</div></li>`
  );
}

/**
 * Markdown as much as a vault needs: headings, list items, paragraphs, and
 * wikilinks as controls.
 *
 * Not a markdown library. What a page written here contains is what the app
 * writes into it plus what a person types, and the shapes below are those. A
 * page that uses more renders as the text it is, which is the honest failure
 * for a preview.
 */
function preview(body: string, links: PageDetail['links']): string {
  const unresolved = new Set(
    links.filter((l) => l.target_kind === 'unresolved').map((l) => l.target_title),
  );
  const out: string[] = [];
  let list: string[] = [];

  const flush = () => {
    if (list.length) out.push(`<ul>${list.join('')}</ul>`);
    list = [];
  };

  for (const line of body.split('\n')) {
    const heading = /^(#{1,4})\s+(.*)$/.exec(line);
    if (heading) {
      flush();
      const level = Math.min(heading[1].length + 1, 5);
      out.push(`<h${level}>${inline(heading[2], unresolved)}</h${level}>`);
      continue;
    }
    const item = /^[-*]\s+(.*)$/.exec(line);
    if (item) {
      list.push(`<li>${inline(item[1], unresolved)}</li>`);
      continue;
    }
    if (!line.trim()) {
      flush();
      continue;
    }
    flush();
    out.push(`<p>${inline(line, unresolved)}</p>`);
  }
  flush();
  return out.join('') || emptyState(COPY.pagesEmptyBody);
}

/** Wikilinks to controls, everything else escaped. */
function inline(text: string, unresolved: Set<string>): string {
  let out = '';
  let rest = text;
  for (;;) {
    const at = rest.indexOf('[[');
    const end = at < 0 ? -1 : rest.indexOf(']]', at);
    if (at < 0 || end < 0) {
      out += esc(rest);
      return out;
    }
    out += esc(rest.slice(0, at));
    const inner = rest.slice(at + 2, end);
    const bar = inner.indexOf('|');
    const target = (bar < 0 ? inner : inner.slice(0, bar)).trim();
    const shown = (bar < 0 ? inner : inner.slice(bar + 1)).trim() || target;
    const missing = unresolved.has(target);
    out +=
      `<button class="wikilink${missing ? ' unresolved' : ''}" ` +
      `data-page-act="${missing ? 'create-link' : 'follow-link'}" ` +
      `data-title="${esc(target)}">${esc(shown)}</button>`;
    rest = rest.slice(end + 2);
  }
}

function backlinksPanel(page: PageDetail): string {
  if (page.backlinks.length === 0) return `<p class="page-note">${COPY.pagesNoBacklinks}</p>`;
  return (
    `<h3>${COPY.pagesBacklinks}</h3><ul class="lib-list">` +
    page.backlinks
      .map(
        (b) =>
          `<li class="lib-row" data-page="${esc(b.page_id)}"><div class="what"><div class="line">` +
          `<span class="title">${esc(b.title)}</span></div>` +
          `<p class="meta">${esc(b.display_text)}</p></div>` +
          `<div class="verbs"><button data-page-act="open">${COPY.pagesOpen}</button></div></li>`,
      )
      .join('') +
    `</ul>`
  );
}

export function pagesHTML(rows: PageRow[], state: PagesState): string {
  if (!state.open) {
    if (rows.length === 0) return emptyState(COPY.pagesEmpty);
    return `<ul class="lib-list">${rows.map(row).join('')}</ul>`;
  }

  const page = state.open;
  if (state.editing) {
    return (
      `<form id="page-edit" class="page-edit">` +
      `<input id="page-name" value="${esc(page.title)}" aria-label="${COPY.pagesTitle}" ` +
      `autocomplete="off" />` +
      `<textarea id="page-text" rows="18" aria-label="${COPY.pagesBody}" ` +
      `spellcheck="false">${esc(page.body)}</textarea>` +
      `<div class="setup-acts"><button type="submit" class="primary">${COPY.pagesSave}</button>` +
      `<span class="note page-file">${COPY.pagesFileNote} ${esc(page.file_path)}</span></div>` +
      `</form>`
    );
  }

  return (
    `<article class="page-read">${preview(page.body, page.links)}</article>` +
    `<p class="page-note page-file">${COPY.pagesFileNote} ${esc(page.file_path)}</p>` +
    backlinksPanel(page)
  );
}

File: crates/tessera-core/src/verbs/pages.rs
//! Pages: the vault, in three reads and two writes.
//!
//! Doc 16 section 3.7's rail item. The explorer lists, the editor reads one and
//! writes it back, and an unresolved link creates the page it names.

use serde::Deserialize;
use serde_json::{Value, json};
use tessera_store::repo;

use crate::core::Core;
use crate::rpc::{Router, RpcError, params};
use crate::verbs::support::*;

pub(super) fn register(r: &mut Router<Core>) {
    r.register("page.list", |core: &mut Core, p| {
        #[derive(Deserialize)]
        struct Limit {
            #[serde(default = "default_pages")]
            limit: i64,
        }
        fn default_pages() -> i64 {
            200
        }
        let p: Limit = params(p)?;
        let pages = repo::list_pages(&core.store, &core.profile_id, p.limit).map_err(store_error)?;
        let rows: Vec<Value> = pages
            .iter()
            .map(|page| {
                json!({
                    "id": page.id,
                    "title": page.title,
                    "file_path": page.file_path,
                    "updated_at": page.updated_at,
                    // Doc 16 section 3.2: a page saved from a card carries the
                    // card's citations, and the explorer says which pages those
                    // are without opening them.
                    "from_card": page.source_card_id.is_some(),
                    "citations_carried": page
                        .citations_carried
                        .as_array()
                        .map(Vec::len)
                        .unwrap_or(0),
                })
            })
            .collect();
        Ok(json!({ "pages": rows }))
    });

    r.register("page.get", |core: &mut Core, p| {
        #[derive(Deserialize)]
        struct Which {
            page_id: String,
        }
        let p: Which = params(p)?;
        let page = repo::read_page(&core.store, &p.page_id)
            .map_err(store_error)?
            .ok_or_else(|| RpcError::core("page_missing", "That page is not in this vault."))?;

        // Doc 16 section 2.1: backlinks are a query over PageLink, never a scan
        // over bodies.
        let backlinks: Vec<Value> = repo::backlinks(&core.store, "page", &p.page_id)
            .map_err(store_error)?
            .into_iter()
            .map(|b| {
                json!({
                    "page_id": b.page_id,
                    "title": b.page_title,
                    "display_text": b.display_text,
                    "position": b.position,
                })
            })
            .collect();

        Ok(json!({
            "id": page.id,
            "title": page.title,
            "body": page.body,
            "file_path": page.file_path,
            "updated_at": page.updated_at,
            "source_card_id": page.source_card_id,
            "citations_carried": page.citations_carried,
            "links": repo::page_links(&core.store, &p.page_id).map_err(store_error)?,
            "backlinks": backlinks,
        }))
    });

    r.register("page.write", |core: &mut Core, p| {
        #[derive(Deserialize)]
        struct Write {
            #[serde(default)]
            page_id: Option<String>,
            title: String,
            #[serde(default)]
            body: String,
        }
        let p: Write = params(p)?;
        let pack_id = core.active_pack_id().map_err(core_error)?;
        let profile_id = core.profile_id.clone();
        let id = crate::vault::write_page(
            &mut core.store,
            &profile_id,
            Some(&pack_id),
            p.page_id.as_deref(),
            &p.title,
            &p.body,
        )
        .map_err(page_error)?;
        let page = repo::read_page(&core.store, &id).map_err(store_error)?;
        Ok(json!({
            "page_id": id,
            "title": page.as_ref().map(|p| p.title.clone()),
            "file_path": page.map(|p| p.file_path),
        }))
    });

    r.register("page.delete", |core: &mut Core, p| {
        #[derive(Deserialize)]
        struct Which {
            page_id: String,
        }
        let p: Which = params(p)?;
        crate::vault::delete_page(&mut core.store, &p.page_id).map_err(store_error)?;
        Ok(json!({ "page_id": p.page_id, "deleted": true }))
    });

    // Doc 16 section 3.1: "Unresolved links create the page on click."
    r.register("page.create_from_link", |core: &mut Core, p| {
        #[derive(Deserialize)]
        struct FromLink {
            title: String,
        }
        let p: FromLink = params(p)?;
        let pack_id = core.active_pack_id().map_err(core_error)?;
        let profile_id = core.profile_id.clone();
        let title = p.title.trim().to_string();
        let body = format!("# {title}\n\n");
        let id = crate::vault::write_page(&mut core.store, &profile_id, Some(&pack_id), None, &title, &body)
            .map_err(page_error)?;
        Ok(json!({ "page_id": id, "title": title }))
    });
}
